### PR TITLE
WIP: fix diloco with shard grap op

### PR DESCRIPTION
**Pull request by samsja**
This PR fixes diloco with shard grap op.

What it does is to handle shared grap op the same way a fully shared is. It, unfortunately, has to be the case for now because FSPD handles SHARD_GRAD_OP like normal full shard expect it does not release the weight after the first all gather. Meaning that the offload pre all gather is not working with shard grap op.

### CODE REVIEW

File: src/zeroband/diloco.py [hunk 1]
```diff
@@ -75,19 +75,16 @@ def __init__(
...
-            self.fsdp_sharding_strategy == ShardingStrategy.FULL_SHARD or self.world_info.local_rank == 0
-        )
-        # if we are not in fully sharded mode only the local rank 0 will have the model on cpu
-        if self.need_to_offload:
-            self._init_offloaded_optimizer(model=model)
-        else:
-            self.outer_optimizer = None
-            self.cpu_model = None
+        if self.fsdp_sharding_strategy not in [ShardingStrategy.FULL_SHARD, ShardingStrategy.SHARD_GRAD_OP]:
+            raise ValueError("Diloco only support FULL_SHARD and SHARD_GRAD_OP")
```
Comment (Jackmin801) [inline src/zeroband/diloco.py:79]: should no shard be supported as well?

**Review by Jackmin801 — APPROVED** [review]
looks good to me. im not sure where the fix is though, didnt you have to summon the full params?

File: src/zeroband/diloco.py (hunk 1 above)
Comment (samsja) [reply src/zeroband/diloco.py:79]: NO_SHARD is failing. I decided not to fix it. I think no shard is handled differently that full shard and shard grad op

**Comment by samsja** [comment]
> looks good to me. I'm not sure where the fix is though, didnt you have to summon the full params?

No, the full-shard one did not work.

I end up basically doing exactly the same as full shard. Since apparently shard grap op handle the param the same way, expect it does not release weight after the all gather

**Pull request merged**